### ISSUE TITLE
Call sites of a type marked with `[CreateNewOnMetadataUpdate]` don't use the updated type on metadata updates

### ISSUE THREAD

**Issue by jeromelaban** [dotnet/roslyn#70604]
**Version Used**: 4.8.0

**Steps to Reproduce**:

1. Create classes (both in the same assembly):
```csharp
[CreateNewOnMetadataUpdate]
public class A
{
  public int F() => 1;
}
public class B
{
  public static void Invoke() => Console.WriteLine(new A().F);
}
```
2. Compile and invoke `B.Invoke()`
3. Make a change to have `F() => 2`
4. Hot Reload
5. Re-run `B.Invoke()`

**Diagnostic Id**: None.

**Expected Behavior**: `2` is printed.

**Actual Behavior**: `1` is printed.

With the current implementation of `[CreateNewOnMetadataUpdate]`, unless I'm missing an option somewhere, call sites that use that type are not re-written to use the new version of a type on change.

This makes source generators that rely on new types quite difficult to write, and in some cases are making some scenarios impossible to do, since user code is involved (and the user can't reference updated types either).

**Comment by Youssef1313** [comment]
cc @davidwengier

**Comment by davidwengier** [comment]
Regardless of the attribute, EnC/HotReload does not make any attempt to update callsites for changed members. In the example posted, if the attribute is removed, then you make an edit that renames `F` to `G`, hot reload and re-run `B.Invoke()` you'll get a `MissingMethodException`.

Also I believe it is expect that user code doesn't need to refer to any type marked with `CreateNewOnMetadataUpdate`, but I'm less sure of that.

**Comment by jeromelaban** [comment]
@davidwengier thanks for the update. 

> Also I believe it is expect that user code doesn't need to refer to any type marked with CreateNewOnMetadataUpdate, but I'm less sure of that.

With regards to `CreateNewOnMetadataUpdate` I have to admit I don't know the original scope of that feature, yet that's not how we are using it for Uno Platform.

We're using it by tagging what are user-defined UI controls with this attribute in order to avoid having to complexify our generated code that uses Lambdas or capture changes, which are not supported edits at this time. With this scenario in mind, it's less important to reload types entirely than hitting rude edits.

Still, as those types are marked with the attribute, the user-defined statements (`new MyType()` statements, specifically) cannot automatically be forwarded to those new generated types unless using some sort of programmatic type-mapping. Even with mapping, still, it requires hoops jumping involving inheritance, interface implementations or `dynamic` for methods that return those replaced types, if methods need to be called.

**Comment by davidwengier** [comment]
whoops, didn't mean to close this, pressed the wrong button

**Comment by tmat** [comment]
This is an expected behavior. When a type marked with `CreateNewOnMetadataUpdate` attribute is changed a new type, with new type identity is emitted (implementation detail: the new type has name suffixed with `#<generation>`, where `<generation>` is the generation ordinal of the update). The original type is not updated.

The common usage pattern of this feature is following:

```C#
// library code:
[assembly: MetadataUpdateHandler(typeof(UpdateHandler))]
interface IPage { ... }

class UpdateHandler
{
  // Update data structures that hold on instances implementing IPage
  static void UpdateApplication(Type[]? updatedTypes) { ... }
  // Remove cached types
  static void ClearCache(Type[]? updatedTypes) { ... }
} 

// generated code:
[CreateNewOnMetadataUpdate]
class Page : IPage { }

// user code:
class C
{
   string Render(IPage page)
   {
      ...
   }
}
```

**Comment by tmat** [comment]
> generated code that uses Lambdas or capture changes, which are not supported edits at this time

We are working on some lambda improvements: https://github.com/dotnet/roslyn/pull/70418.

With this change, methods that use _ephemeral_ lambdas will be allowed to make changes to the lambdas.
_Ephemeral_ means a lambda that is created, used and discarded before the user has the chance to update its body.

For example, a lambda used as an event handler is not _ephemeral_ since it is expected that the code that subscribes to the event runs ones, registering the handler, and then whenever the handler is invoked the latest version of the lambda body executes until the handler is unsubscribed.

On the other hand, lambdas used in page rendering code generated by Razor are _ephemeral_. The render method creates new delegates for lambdas it contains every time it's called. These lambdas are passed around just for callbacks that happen during the same render invocation and then discarded.

**Comment by jeromelaban** [comment]
Thanks for the update @tmat! 

Considering the scenario that you provided it indeed makes sense that it was implemented this way. Would changing call sites be a consideration? I'm not too sure if that would make sense performance wise (as non-modified documents may be impacted by a change made elsewhere).

Now, as parts of lambdas are in the works, it makes this request less interesting for sure, but it would definitely improve our scenario as we could take into account that particular restriction to generate code in debug.

What would happen if the modified lambdas are non-ephemeral? Would it fail as compile time as a rude edit? Or would the older version of the display class be kept and not modified? (which would effectively keep the old code running)

**Comment by tmat** [comment]
> Considering the scenario that you provided it indeed makes sense that it was implemented this way. Would changing call sites be a consideration? 

No, we can't update the call sites. The updated type is entirely different from the original one. The only thing that ties them together is an attribute we emit on the new one that points to the original one. Specifically, the new type is not a subtype of the previous type. We can't change the type of an existing object instance and we can't assign the new type to variables typed to the old one.

> What would happen if the modified lambdas are non-ephemeral? Would it fail as compile time as a rude edit? Or would the older version of the display class be kept and not modified? (which would effectively keep the old code running)

If the lambda update changes the closure in an incompatible way the old version of the lambda will be updated to throw an exception when invoked. So as long as the old lambda is not invoked it's all good.

**Comment by NMSAzulX** [comment]
@tmat Can hot reload be used while the program is running? When I run the program (not f5), the value of MetadataUpdater.isSupported is false. I hope to dynamically generate class A and update the old class A while the program is running. :)
